### PR TITLE
add MultiMap interface with GoDocs

### PR DESCRIPTION
**Pull request by furkansenharputlu**
#164

### CODE REVIEW

**Comment by sancar** [comment]
`delete` seems missing

File: core/multi_map.go [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+// MultiMap is a specialized map whose keys can be associated with multiple values.
+type MultiMap interface {
+	// IDistributedObject is the base interface for all distributed objects.
+	IDistributedObject
+
+	// Stores a key-value pair in the multi-map.
+	// It returns true if size of the multi-map is increased, false if the multi-map
+	// already contains the key-value pair.
+	Put(key interface{}, value interface{}) (increased bool, err error)
```
Comment (SaitTalhaNisanci) [inline core/multi_map.go]: I think it would be better to use a more descriptive return name here, maybe "*changed*" ?

File: core/multi_map.go [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+	Get(key interface{}) (values []interface{}, err error)
+
+	// Remove removes an association of the specified value with the specified key. Calling this method does not affect
+	// other values associated with the same key.
+	// It returns true if the value was detached from the specified key, false if it was not.
+	Remove(key interface{}, value interface{}) (removed bool, err error)
+
+	// RemoveAll detaches all values from the specified key.
+	// It returns a slice of old values that were associated with this key prior to this method call.
+	RemoveAll(key interface{}) (oldValues []interface{})
```
Comment (SaitTalhaNisanci) [inline core/multi_map.go]: This method should also return an error.

File: core/multi_map.go [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+	ContainsEntry(key interface{}, value interface{}) (found bool, err error)
+
+	// Clear removes all entries from this multi-map.
+	Clear() (err error)
+
+	// Size returns the total number of values in this multi-map.
+	Size() (size int32, err error)
+
+	// ValueCount returns the number of values associated with the specified key.
+	ValueCount(key interface{}) (valueCount int32)
```
Comment (SaitTalhaNisanci) [inline core/multi_map.go]: This method should also return an error

**Comment by furkansenharputlu** [comment]
@sancar What do you mean by `delete`, a method ? I cannot see it in Java client

File: core/multi_map.go (hunk 1 above)
Comment (furkansenharputlu) [reply core/multi_map.go]: I think "*increased*" is descriptive enough.

**Comment by sancar** [comment]
@furkansenharputlu 
It seems `delete` is recently added(3.10). Here:
https://github.com/hazelcast/hazelcast/blob/master/hazelcast-client/src/main/java/com/hazelcast/client/proxy/ClientMultiMapProxy.java#L153

File: core/multi_map.go (hunk 1 above)
Comment (asimarslan) [reply core/multi_map.go]: `changed` is not correct as it includes decrease.

We can use something like `mapSizeIncreased`

File: core/multi_map.go [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+	// Stores a key-value pair in the multi-map.
+	// It returns true if size of the multi-map is increased, false if the multi-map
+	// already contains the key-value pair.
+	Put(key interface{}, value interface{}) (increased bool, err error)
+
+	// Get returns a slice of values associated with the specified key.
+	Get(key interface{}) (values []interface{}, err error)
+
+	// Remove removes an association of the specified value with the specified key. Calling this method does not affect
```
Comment (asimarslan) [inline core/multi_map.go]: First `Remove` is not needed.  
"Removes and association ... " is better

**Pull request merged**